# **<span style="color:red">Taller: Introducción a la programación en Python</span>**

#  **<span style="color:blue">Método de newton-Raphson</span>**

El método de *Newton* (conocido también como el método de **Newton-Raphson**) es un algoritmo para encontrar aproximaciones de los ceros o raíces de una función real. 

También puede ser usado para encontrar el máximo o mínimo de una función, encontrando los ceros de su primera derivada.

El método de *Newton-Raphson* es uno de los métodos más utilizados.

Un **algoritmo** para este método método es el siguiente:

Para determinar una raiz de $f(x)=0$, dado un valor $x_0$ razonablemente próximo a la raiz:

Repetir:\
$\ \ \ \ $ Calcular $x_1=x_0-\displaystyle\frac{f(x_0)}{f^\prime(x_0)}$\
$\ \ \ \ $ Hacer $x_0=x_1$    
$\ \ \ \ $ Si $\left |x_0-x_1 \right | <$ valor de tolerancia o $f(x_0)=0$ o n = número de iteraciones\
$\ \ \ \ $ $\ \ \ \ $ Fin

**<span style="color:blue">Ejemplo:</span>**

Aproxime una solución de $f(x)=e^{x-1}-5x^3$ con el método de *Newton-Rapshon*. Empezando con $x_0=1$ y con n=4 iteraciones.

**<span style="color:blue">Solución:</span>**

Tenemos $f(x)=e^{x-1}-5x^3$ $\implies$ $f^\prime(x)=e^{x-1}-15x^2$.

*Iteración 1*:\
$x_1=x_0-\displaystyle\frac{f(x_0)}{f^\prime(x_0)}=1-\displaystyle\frac{e^{1-1}-5(1)^3}{e^{1-1}-15(1)^2}=1-\displaystyle\frac{-4}{-14}=0.7142857143$

Hacemos $x_0 = x_1 = 0.7142857143$

Iteración 2:\
$x_1=0.7142857143-\displaystyle\frac{e^{0.7142857143-1}-5(0.7142857143)^3}{e^{0.7142857143-1}-15(0.7142857143)^2}=0.5591502915$

Hacemos $x_0 = x_1 = 0.5591502915$

Iteración 3:\
$x_1=0.5591502915-\displaystyle\frac{e^{0.5591502915-1}-5(0.5591502915)^3}{e^{0.5591502915-1}-15(0.5591502915)^2}=0.5021592918$

Hacemos $x_0 = x_1 = 0.5021592918$

Iteración 4:\
$x_1=0.5021592918-\displaystyle\frac{e^{0.5021592918-1}-5(0.5021592918)^3}{e^{0.5021592918-1}-15(0.5021592918)^2}=0.4941927850$

Hacemos $x_0 = x_1 = 0.4941927850$

Con $n=4$ iteraciones, vemos que una solución aproximada de: $f(x)=e^{x-1}-5x^3$ es: 0.4941927850.

**<span style="color:blue">Ejercicio 1:</span>** 

Escriba un programa que resuelva ecuaciones no lineales utilizando el método de *Newton-Raphson*. El programa debe pedir al usuario: 
- la función $f(x)$ de la cual quiere obtener un cero o una raiz
- la derivada de $f(x)$: $f^\prime(x)$
- el valor inicial: $x_0$
- el número máximo de iteraciones
- un valor de tolerancia

Imprimir la solución, y el número de iteraciones que utilizó.

*Nota 1*: su programa debe preguntar al usuario si desea volver a utilizar el método de *Newton-Raphson*. Si su respuesta es *sí*, debe repetir los pasos anteriores. Si su respuesta es *no*, finalizar el programa.  

*Nota 2*: debe manejar las excepciones a la hora de pedir el valor inicial, el número de iteraciones y el valor de tolerancia.

*Nota 3:* si al preguntar al usuario si desea volver a utilizar el método de *Newton-Raphson*, el usuario escribe algo distinto a *sí* o *no*, mandar un mensaje diciendo que las respuestas posibles son *sí* o *no*, y darle la opción de insertar una respueta nuevamente hasta que escriba una de los dos respuestas posibles.

**<span style="color:blue">Solución:</span>** 

In [17]:
# Solución
print("="*78)
print(" "*30+"Método de Newton")
print("="*78)
print("Recordemos que en el método de Newton necesitamos la función f(x), la derivada")
print("de f(x): f'(x) y un valor inicial x0")
print(" ")

import math 

# Defimos la función para valorar datos de entrada (números)
def leer_número(mensaje):
    while True:
        try:
            entrada = float(input(mensaje))
            return(entrada)
        except ValueError:
            print(" ")
            print("¡Error! Debe insertar un número. \n")
        else:
            break
            
# Definimos la función Newton
def Newton(x0,n,tolerancia):
    contar = 0
    parar = False
    while not parar:
        contar += 1
        x1 = x0-((f(x0))/(fprima(x0)))
        if (contar==n) or (f(x0)==0) or (abs(x0-x1)<tolerancia):
            parar = True
        x0 = x1
    print("La solución es: " + str(x0))
    print("Iteraciones:",contar)
    
salir = False
while not salir:
    # Solicitar al usuario la función f(x)
    funcion = input("Ingrese la función (en código python) de la cuál quiere encontrar una solución:\n")
    def f(x):
        return(eval(funcion))
    # Solicitar al usuario la derivada de la función f(x): f'(x)
    derivada = input("Ingrese la derivada de la función (en código python) de la cuál quiere encontrar una solución:\n")
    def fprima(x):
        return(eval(derivada))
    x0 = leer_número("Ingrese el valor incial de x0: ")
    n = leer_número("Ingrese el número máximo de iteraciones: ")
    tolerancia = leer_número("Ingrese el valor de tolerancia: ")
    print(" ")
    Newton(x0,n,tolerancia)
    
    print("\n \n ¿Desea utilizar el método de Newton nuevamente?\n ")
    while True:
        pregunta = input("(sí/no): ")
        if (pregunta=="No") or (pregunta=="no") or (pregunta=="NO"):
            print("\n Ok, ¡Qué os vaya bien!")
            salir = True
            break
        elif (pregunta=="Sí") or (pregunta=="sí") or (pregunta=="SÍ") or (pregunta=="Si") or (pregunta=="si") or (pregunta=="SI"):
            break
        else:
            print("\n Las respuestas posibles son sí o no \n")

                              Método de Newton
Recordemos que en el método de Newton necesitamos la función f(x), la derivada
de f(x): f'(x) y un valor inicial x0
 
Ingrese la función (en código python) de la cuál quiere encontrar una solución:
x**x-100
Ingrese la derivada de la función (en código python) de la cuál quiere encontrar una solución:
x**x*(1+math.log(x))
Ingrese el valor incial de x0: 5
Ingrese el número máximo de iteraciones: 30
Ingrese el valor de tolerancia: 0.0001
 
La solución es: 3.597285023540423
Iteraciones: 8

 
 ¿Desea utilizar el método de Newton nuevamente?
 
(sí/no): no

 Ok, ¡Qué os vaya bien!


Aproxime una solución de $f(x)=e^{x-1}-5x^3$ con el método de *Newton-Rapshon*. Empezando con $x_0=1$ y con n=4 iteraciones.

**<span style="color:blue">Ejercicio 2:</span>**
    
Usando su programa verifique la solución del ejemplo.

**<span style="color:blue">Ejercicio 3:</span>**

Usando su programa. Aproxime una solución de $x^x=100$ con el método de *Newton-Rapshon*. Elija un valor para $x_0$, un número de iteraciones y una tolerancia.

Hint: La derivada de $f(x)=x^x-100$ es $f^\prime(x)=x^x(1+ln(x))$.

**<span style="color:blue">Solución:</span>**

Una solución aproximada para $x^x=100$ utilizando el método *Newton-Rapshon* es: 3.597285023540423